### DISCUSSION TITLE
[CUDA] Questions about memory release

### DISCUSSION THREAD

**Discussion by fffrog** [microsoft/onnxruntime#13399] [Other Q&A] [answered]
The ```temp_mask_buffer``` variable will execute its destructor because it goes out of scope, and then call cudaFree to release the memory it refers to; since the cuda kernel is executed asynchronously relative to the cpu, the cuda kernel ```DropoutKernel``` may not be executed when the memory is released.

My understanding may not be correct, could someone take the time to explain it? Thanks a lot.

https://github.com/microsoft/onnxruntime/blob/a8701c2a592ee017e21a984df0269bfd268e15ae/onnxruntime/core/providers/cuda/nn/dropout.cc#L110

https://github.com/microsoft/onnxruntime/blob/a8701c2a592ee017e21a984df0269bfd268e15ae/onnxruntime/core/providers/cuda/nn/dropout_impl.cu#L32

**Comment by pranavsharma** [comment, marked as answer]
The temp buffer is allocated using an arena whose memory is not freed. Hence async execution is not affected.
